src/Main.hs

```diff
@@ -176,50 +176,152 @@ defaultStore =
+  return $ [(Invariants invariants [contract])]
+           ++ (splitCase name True contract iface [] iffs' Nothing stateUpdates postcs [contract])
+
+mkEnv :: Id -> Store -> [Decl]-> Env
+mkEnv contract store decls = (fromMaybe mempty (Map.lookup contract store), store, abiVars)
+ where
+   abiVars = Map.fromList $ map (\(Decl typ var) -> (var, metaType typ)) decls
+
+-- split case into pass and fail case
+splitCase :: String -> Bool -> String -> Interface -> [Exp Bool] -> [Exp Bool] -> Maybe ReturnExp
+          -> Map String [Either StorageLocation StorageUpdate] -> [Exp Bool] -> [String] -> [Behaviour]
+splitCase name creates contract iface ifs [] ret storage postcs contracts =
+  [ Behaviour name Pass creates contract iface (mconcat ifs) (mconcat postcs) contracts storage ret ]
+splitCase name creates contract iface ifs iffs ret storage postcs contracts =
+  [ Behaviour name Pass creates contract iface (mconcat (ifs <> iffs)) (mconcat postcs) contracts storage ret,
+    Behaviour name Fail creates contract iface (And (mconcat ifs) (Neg (mconcat iffs))) (mconcat postcs) contracts storage Nothing ]
+
+-- extracts bounds on Integer values in storage, returns Iff or Exp Bool
+-- representations for use in either pre or post conditions
+getStorageBounds :: Env -> ([Exp Bool], [IffH])
+getStorageBounds (ours, _, _) =
+  unzip $ catMaybes $ fmap getBound $ Map.toList ours
+  where
+    getBound :: (Id, SlotType) -> Maybe (Exp Bool, IffH)
+    getBound (id, (StorageValue typ)) = case metaType typ of
+      Integer -> Just $ (bound typ (IntVar id), IffIn nowhere typ [EntryExp nowhere id []])
+      _ -> Nothing
+    getBound (_, _) = Nothing
+
+-- extract a list of iff headers from the size of the types in a list of calldata declarations
+getCallDataBounds :: [Decl] -> [IffH]
+getCallDataBounds decls =
+  join $
+    fmap
+      ( \(Decl typ id) -> case metaType typ of
+          Integer -> [IffIn nowhere typ [EntryExp nowhere id []]]
+          _ -> []
+      )
+      decls
+
+-- ensures that key types match value types in an assign
+checkAssign :: Env -> Assign -> Err (Id, [StorageUpdate])
+checkAssign env (AssignVal (StorageVar (StorageValue typ) id) expr)
+  = case metaType typ of
+    Integer -> do
+      val <- checkInt env expr
+      return (id, [IntUpdate (DirectInt id) val])
+    Boolean -> do
+      val <- checkBool env expr
+      return (id, [BoolUpdate (DirectBool id) val])
+    ByteStr -> do
+      val <- checkBytes env expr
+      return (id, [BytesUpdate (DirectBytes id) val])
+checkAssign env (AssignMany (StorageVar (StorageMapping (keyType :| _) valType) id) defns)
+  = do updates <- mapM (checkDefn env keyType valType id) defns
+       return $ (id, updates)
+checkAssign env (AssignVal (StorageVar (StorageMapping _ _) _) _)
+  = Bad (nowhere, "Cannot assign a single expression to a composite type")
+checkAssign env (AssignMany (StorageVar (StorageValue _) _) _)
+  = Bad (nowhere, "Cannot assign multiple values to an atomic type")
```

Comment:
Would it make sense to use a `Defn` with e.g. an array type?

Reply:
Hmm, I don't think we want to give arrays an index like that. In general, there's a few things missing to deal with arrays right now, I think they are not really a priority atm.

I'm thinking that the way to deal with them would be to add some array flavored expressions to `Expr`, such as `List [Expr]`, and corresponding zooming / lookup operations